src/economy/waterway.cc

```diff
@@ -94,13 +94,12 @@ void Waterway::remove_worker(Worker& w) {
 
 void Waterway::request_ferry(EditorGameBase& egbase) {
 	FerryFleet* fleet = new FerryFleet(get_owner());
-	fleet->request_ferry(egbase, this);
-	fleet->init(egbase);
+	fleet->init(egbase, this);
 }
 
 void Waterway::assign_carrier(Carrier& c, uint8_t) {
-	if (ferry_) {
-		ferry_->set_location(nullptr);
+	if (ferry_.get(owner().egbase())) {
+		ferry_.get(owner().egbase())->set_location(nullptr);
```

Comment:
Duplicate calls to get() should be made more efficient by caching the return value, e.g.
```suggestion
	if (Ferry* f = ferry_.get(owner().egbase())) {
		f->set_location(nullptr);
```
There are more places in the diff